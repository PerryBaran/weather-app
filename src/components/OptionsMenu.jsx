import React, { useState } from "react";
import PropTypes from "prop-types";
import "../styles/OptionsMenu.css";
import useUnitContext from "../hooks/useUnitContext";
import menu from "../images/menu.png";

function OptionsMenu({ handleSearch }) {
  const { units, setUnits } = useUnitContext();
  const [visisble, setVisible] = useState(false);
  const [selectedUnits, setSelectedUnits] = useState(units);

  const handleHamburger = () => {
    setVisible((prev) => !prev);
  };

  const changeHandler = (e) => {
    setSelectedUnits(e.target.value);
  };

  const handleApply = (e) => {
    e.preventDefault();
    setUnits(selectedUnits);
    handleSearch(undefined, selectedUnits);
    setVisible(false);
  };

  return (
    <>
      <button
        type="button"
        className="options-menu-button"
        onClick={handleHamburger}
      >
        <img src={menu} alt="options menu" />
      </button>
      <div className="options-menu-form__container">
        <form
          className={`options-menu-form ${!visisble && "hidden"}`}
          name="options menu form"
        >
          <h2>Options</h2>
          <div className="options-menu-form__units">
            <h3>Units:</h3>
            <label htmlFor="standardRadio" className="options-menu-form__label">
              Scientific
              <input
                id="standardRadio"
                type="radio"
                value="standard"
                name="units"
                checked={selectedUnits === "standard"}
                onChange={changeHandler}
              />
            </label>
            <label htmlFor="metricRadio" className="options-menu-form__label">
              Metric
              <input
                id="metricRadio"
                type="radio"
                value="metric"
                name="units"
                checked={selectedUnits === "metric"}
                onChange={changeHandler}
              />
            </label>
            <label htmlFor="imperialRadio" className="options-menu-form__label">
              Imperial
              <input
                id="imperialRadio"
                type="radio"
                value="imperial"
                name="units"
                checked={selectedUnits === "imperial"}
                onChange={changeHandler}
              />
            </label>
          </div>
          <button
            type="submit"
            onClick={handleApply}
            className="options-menu-form__apply"
          >
            Apply
          </button>
        </form>
      </div>
    </>
  );
}

OptionsMenu.propTypes = {
  handleSearch: PropTypes.func.isRequired,
};

export default OptionsMenu;
