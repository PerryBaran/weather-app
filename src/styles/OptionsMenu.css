
.options-menu {
  width: 200px;  
  height: 100%;
  margin-left: 20px;
  text-align: center;
  font-weight: 300;
  color: white;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.options-menu-button {
  all: unset;
  cursor: pointer;
}

.options-menu-button > img {
  aspect-ratio: 1;
  width: 40px;
}

.options-menu-form__container {
  position: absolute;
  top: var(--top-bar-height);
  right: 0;
  overflow: hidden;
}

.options-menu-form {
  padding: 10px;
  background-color: var(--medium-blue);
  position: relative;
  top: 0;
  transition-duration: 500ms;
  border-bottom-left-radius: 5px;
}

.hidden {
  top: -200px;
}

.options-menu-form h3 {
  color: var(--dark-blue);
  font-weight: 500;
  font-size: 1.6rem;
  text-decoration: underline;
}

.options-menu-form__units {
  display: flex;
  align-items: center;
  margin: 10px 0;
}

.options-menu-form__units h4 {
  font-size: 1.2rem;
  margin-right: 5px;
}

.options-menu-form__radio {
  display: none;
}

.options-menu-form__label:has(.options-menu-form__radio:checked) {
  background-color: var(--light-blue);
  color: var(--dark-blue);
}

.options-menu-form__label {
  border: 1px solid white;
  cursor: pointer;
  padding: 5px;
  color: var(--almost-white);
  background-color: var(--dark-blue);
}

