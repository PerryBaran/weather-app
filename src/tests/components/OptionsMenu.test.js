import React, { createContext, useContext } from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import * as useUnitContext from "../../hooks/useUnitContext";
import OptionsMenu from "../../components/OptionsMenu";

const findRadio = (radios, value) => {
  return radios.find((radio) => radio.value === value);
};

describe("OptionsMenu", () => {
  const validProps = {
    setVisible: jest.fn(),
  };

  test("snapshot", () => {
    const { asFragment } = render(
      <OptionsMenu setVisible={validProps.setVisible} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(() => {
      render(<OptionsMenu setVisible={validProps.setVisible} />);
    });

    test("all fields correctly rendered", () => {
      const { getByText, getAllByRole, getByRole } = screen;
      const radios = getAllByRole("radio");
      const standard = findRadio(radios, "standard");
      const metric = findRadio(radios, "metric");
      const imperial = findRadio(radios, "imperial");
      const apply = getByRole("button");

      expect(getByText(/options/i)).toBeInstanceOf(HTMLHeadingElement);
      expect(getByText(/units:/i)).toBeInstanceOf(HTMLHeadingElement);
      expect(radios).toHaveLength(3);
      expect(standard).toHaveAttribute("id", "standardRadio");
      expect(getByText(/scientific/i)).toHaveAttribute("for", "standardRadio");
      expect(metric).toHaveAttribute("id", "metricRadio");
      expect(getByText(/metric/i)).toHaveAttribute("for", "metricRadio");
      expect(imperial).toHaveAttribute("id", "imperialRadio");
      expect(getByText(/imperial/i)).toHaveAttribute("for", "imperialRadio");
      expect(apply).toHaveAttribute("type", "submit");
      expect(apply).toHaveTextContent(/apply/i);
    });

    test("apply button calls setVisible", () => {
      const { getByRole } = screen;
      const apply = getByRole("button");

      expect(validProps.setVisible).toBeCalledTimes(0);
      fireEvent.click(apply);
      expect(validProps.setVisible).toBeCalledWith(false);
    });
  });

  describe("mocked useContext", () => {
    let mockSetUnits;

    beforeEach(() => {
      mockSetUnits = jest.fn();
      const mockUnitContext = createContext({
        units: "metric",
        setUnits: mockSetUnits,
      });
      jest
        .spyOn(useUnitContext, "default")
        .mockImplementation(() => useContext(mockUnitContext));
      render(<OptionsMenu setVisible={validProps.setVisible} />);
    });

    test("selecting units", () => {
      const { getByRole, getAllByRole } = screen;
      const [standard, metric, imperial] = getAllByRole("radio");
      const apply = getByRole("button");

      fireEvent.click(standard);
      expect(standard).toBeChecked();
      expect(metric).not.toBeChecked();
      expect(imperial).not.toBeChecked();
      fireEvent.click(apply);
      expect(mockSetUnits).toBeCalledWith(standard.value);
      fireEvent.click(metric);
      expect(standard).not.toBeChecked();
      expect(metric).toBeChecked();
      expect(imperial).not.toBeChecked();
      fireEvent.click(apply);
      expect(mockSetUnits).toBeCalledWith(metric.value);
      fireEvent.click(imperial);
      expect(standard).not.toBeChecked();
      expect(metric).not.toBeChecked();
      expect(imperial).toBeChecked();
      fireEvent.click(apply);
      expect(mockSetUnits).toBeCalledWith(imperial.value);
    });
  });
});
