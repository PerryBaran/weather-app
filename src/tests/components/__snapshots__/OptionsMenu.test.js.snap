// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsMenu snapshot 1`] = `
<DocumentFragment>
  <button
    class="options-menu-button"
    type="button"
  >
    <img
      alt="options menu"
      src="menu.png"
    />
  </button>
  <div
    class="options-menu-form__container"
  >
    <form
      class="options-menu-form hidden"
      name="options menu form"
    >
      <h2>
        Options
      </h2>
      <div
        class="options-menu-form__units"
      >
        <h3>
          Units:
        </h3>
        <label
          class="options-menu-form__label"
          for="standardRadio"
        >
          Scientific
          <input
            checked=""
            id="standardRadio"
            name="units"
            type="radio"
            value="standard"
          />
        </label>
        <label
          class="options-menu-form__label"
          for="metricRadio"
        >
          Metric
          <input
            id="metricRadio"
            name="units"
            type="radio"
            value="metric"
          />
        </label>
        <label
          class="options-menu-form__label"
          for="imperialRadio"
        >
          Imperial
          <input
            id="imperialRadio"
            name="units"
            type="radio"
            value="imperial"
          />
        </label>
      </div>
      <button
        class="options-menu-form__apply"
        type="submit"
      >
        Apply
      </button>
    </form>
  </div>
</DocumentFragment>
`;
